src/openrct2-ui/windows/Footpath.cpp

```diff
@@ -164,749 +137,847 @@ static constexpr const uint8_t DefaultPathSlope[] = {
+        WindowInitScrollWidgets(this);
+        window_push_others_right(this);
+        show_gridlines();
+
+        tool_cancel();
+        _footpathConstructionMode = PATH_CONSTRUCTION_MODE_LAND;
+        tool_set(this, WIDX_CONSTRUCT_ON_LAND, Tool::PathDown);
+        input_set_flag(INPUT_FLAG_6, true);
+        _footpathErrorOccured = false;
+        SetEnabledAndPressedWidgets();
     }
 
-    // Check if window is already open
-    rct_window* window = window_bring_to_front_by_class(WC_FOOTPATH);
-    if (window != nullptr)
+    void OnClose() override
     {
-        return window;
+        footpath_provisional_update();
+        viewport_set_visibility(0);
+        map_invalidate_map_selection_tiles();
+        gMapSelectFlags &= ~MAP_SELECT_FLAG_ENABLE_CONSTRUCT;
+        window_invalidate_by_class(WC_TOP_TOOLBAR);
+        hide_gridlines();
     }
 
-    window = WindowCreate(ScreenCoordsXY(0, 29), WW, WH, &window_footpath_events, WC_FOOTPATH, 0);
-    window->widgets = window_footpath_widgets;
-    window->enabled_widgets = (1ULL << WIDX_CLOSE) | (1ULL << WIDX_FOOTPATH_TYPE) | (1ULL << WIDX_QUEUELINE_TYPE)
-        | (1ULL << WIDX_RAILINGS_TYPE) | (1ULL << WIDX_DIRECTION_NW) | (1ULL << WIDX_DIRECTION_NE) | (1ULL << WIDX_DIRECTION_SW)
-        | (1ULL << WIDX_DIRECTION_SE) | (1ULL << WIDX_SLOPEDOWN) | (1ULL << WIDX_LEVEL) | (1ULL << WIDX_SLOPEUP)
-        | (1ULL << WIDX_CONSTRUCT) | (1ULL << WIDX_REMOVE) | (1ULL << WIDX_CONSTRUCT_ON_LAND)
-        | (1ULL << WIDX_CONSTRUCT_BRIDGE_OR_TUNNEL);
-
-    WindowInitScrollWidgets(window);
-    window_push_others_right(window);
-    show_gridlines();
-
-    tool_cancel();
-    _footpathConstructionMode = PATH_CONSTRUCTION_MODE_LAND;
-    tool_set(window, WIDX_CONSTRUCT_ON_LAND, Tool::PathDown);
-    input_set_flag(INPUT_FLAG_6, true);
-    _footpathErrorOccured = false;
-    WindowFootpathSetEnabledAndPressedWidgets();
+    void OnUpdate() override
+    {
+        widget_invalidate(this, WIDX_CONSTRUCT);
+        UpdateProvisionalPathForBridgeMode();
 
-    return window;
-}
+        // #2502: The camera might have changed rotation, so we need to update which directional buttons are pressed
+        uint8_t currentRotation = get_current_rotation();
+        if (_lastUpdatedCameraRotation != currentRotation)
+        {
+            _lastUpdatedCameraRotation = currentRotation;
+            SetEnabledAndPressedWidgets();
+        }
 
-/**
- *
- *  rct2: 0x006A852F
- */
-static void WindowFootpathClose(rct_window* w)
-{
-    footpath_provisional_update();
-    viewport_set_visibility(0);
-    map_invalidate_map_selection_tiles();
-    gMapSelectFlags &= ~MAP_SELECT_FLAG_ENABLE_CONSTRUCT;
-    window_invalidate_by_class(WC_TOP_TOOLBAR);
-    hide_gridlines();
-}
+        // Check tool
+        if (_footpathConstructionMode == PATH_CONSTRUCTION_MODE_LAND)
+        {
+            if (!(input_test_flag(INPUT_FLAG_TOOL_ACTIVE)))
+            {
+                window_close(this);
```

Comment:
Use the member function `Close` instead (it just calls `window_close` but is neater). This comment applies accross this file in all the instances of window_close